### ISSUE TITLE
Contrastive learning + GNN for predicting NPV of patents

### ISSUE THREAD

**Issue by kristinlindquist** [biosymbolics/pipeline#65]
https://www.pnas.org/doi/10.1073/pnas.2220778120
<img width="1237" alt="Screenshot 2023-07-21 at 2 33 46 PM" src="https://github.com/biosymbolics/pipeline/assets/9382486/48435fbf-06c2-45af-bc62-9f09fa3a247f">
<img width="849" alt="Screenshot 2023-07-21 at 2 34 26 PM" src="https://github.com/biosymbolics/pipeline/assets/9382486/4321be6d-4f7f-497d-b146-f239386c08b2">

My question to Claude:

> Is it possible to have a composite traditional contrastive DNN with a GNN subcomponent? In particular, I want to train a DNN with patent data (tagged compounds and indications) using "did it get into phase X clinical trials" and "did it get regulatory approval" as the supervised signal for "good", maximizing the difference between compounds that were "successful" vs not, but I want to break down the meaning of a patent into "this is a vote in favor of X being true" where X is "compoundX has mechanismX'", "mechanismX' affects pathway X which in turn successfully intervenes on diseaseX''". Thus there would be GNN nodes, in a sense, that were able to learn the relationships to the multi-factorial nature of an intervention's success, whereas the overall contrastive DNN would make sense of it overall (since Compound -> Mechanism -> Pathway -> Disease data is both sparse and biased).

Claude's answer:

> Yes, it is absolutely possible to have a composite neural network model that combines traditional contrastive DNN components with graph neural network (GNN) subcomponents. Here is one way this could be implemented:
> 
> The overall model architecture would be a contrastive DNN like a Siamese or triplet network, taking in patent data vectors as input and optimized using your "success" vs "not success" supervised labels.
> This DNN could encode the patent data into hidden representations that summarize the key compound, mechanism, pathway, disease relationships.
> Within the DNN would be a GNN submodule that explicitly models the compound->mechanism->pathway->disease relationships as a graph.
> The GNN would include node layers representing compounds, mechanisms, pathways, diseases. Edge layers would model their relationships.
> The GNN outputs could feed into the DNN hidden layers, providing relational knowledge.
> The overall contrastive loss and backprop would train both the DNN and GNN components jointly.
> This allows the GNN to "vote" on relevant relationships while the DNN models overall success patterns.
> During training, the GNN connections can be selectively masked to deal with missing data. Inference could focus on most probable relationships. With joint training, the GNN would reciprocally gain domain knowledge from the DNN as well. This composite approach allows modeling both overall success patterns as well as drilling into the explanatory factors behind a prediction.

```
# Contrastive DNN 
dnn_input = Input(shape=(patent_data_dim,))
dnn_x = Dense(512)(dnn_input)
dnn_x = BatchNormalization()(dnn_x)  
dnn_x = Activation('relu')(dnn_x)
dnn_x = Model(inputs=dnn_input, outputs=dnn_x)

# GNN component
gnn_input = Input(shape=(num_compounds, num_features))
gnn_x = GraphConv(32)([gnn_input, adjacency_matrix])
gnn_x = BatchNormalization()(gnn_x)
gnn_x = Activation('relu')(gnn_x)
gnn_x = GraphConv(64)([gnn_x, adjacency_matrix])
gnn_x = BatchNormalization()(gnn_x)
gnn_x = Activation('relu')(gnn_x)
gnn_x = GlobalAvgPooling()(gnn_x)
gnn_x = Model(inputs=gnn_input, outputs=gnn_x)

# Combine DNN and GNN
combined_input = Input(shape=(patent_data_dim,))
dnn_emb = dnn_x(combined_input)
gnn_emb = gnn_x(gnn_input) 

x = Concatenate()([dnn_emb, gnn_emb])
x = Dense(128)(x)

# Success vs non-success output
output = Dense(1, activation='sigmoid')(x) 

model = Model(inputs=[combined_input, gnn_input], outputs=output)

model.compile(loss='binary_crossentropy', optimizer='adam')
model.fit(data)
```
